myia/utils/misc.py

```diff
@@ -415,3 +415,13 @@ def eprint(*things):
 def is_dataclass_type(cls):
     """Returns whether cls is a dataclass."""
     return isinstance(cls, type) and hasattr(cls, '__dataclass_fields__')
+
+
+def as_frozen(x):
+    """Return an immutable representation for x."""
+    if isinstance(x, dict):
+        return tuple((k, as_frozen(v)) for k, v in x.items())
```

Comment:
This should also sort the returned items otherwise two calls with `{'x': 1, 'y': 2}`  and `{'y': 2, 'x': 1}` would be unequal.

Reply:
True that.